modeshape-jcr/src/main/java/org/modeshape/jcr/value/binary/DatabaseBinaryStore.java

```diff
@@ -173,14 +174,29 @@ public BinaryValue storeValue( InputStream stream, final boolean markAsUnused )
                         database.restoreContent(connection, Collections.singletonList(key));
                     }
                 } else {
-                    // store the content
-                    database.insertContent(key, temp.getStream(), temp.getSize(), connection);
-                    if (markAsUnused) {
-                        database.markUnused(Collections.singletonList(key), connection);
+                    try (InputStream is = temp.getStream()) {
+                        // store the content
+                        database.insertContent(key, is, temp.getSize(), connection);
+                        if (markAsUnused) {
+                            database.markUnused(Collections.singletonList(key), connection);
+                        }
                     }
                 }
                 return new StoredBinaryValue(DatabaseBinaryStore.this, key, temp.getSize());
             });
+        } catch (BinaryStoreException e) {
+            if (e.getCause() instanceof SQLException) {
+                // under certain conditions - e.g. in a cluster - someone else may have already inserted the binary
+                // so try reading again
+                 return dbCall(connection -> {
+                    if (database.contentExists(key, !markAsUnused, connection)) {
+                        return new StoredBinaryValue(DatabaseBinaryStore.this, key, temp.getSize());        
+                    }
+                    // nothing there, so rethrow the original exception
+                    throw e;
+                });
+            }
+            throw e;
```

Comment:
If you look at the [dbCall](https://github.com/ModeShape/modeshape/pull/1645/files#diff-d1dc99bb1634087ce64bcb5a66ecc4e4L329) method, that's the only place where the exception can come from since that's where the transaction would be committed.

In other words, the only way to get a `SQLException` in this method (storeValue) would be to change the signature of `dbCall` which is used by several other places that rely only on `BinaryStoreException`

Either that or have a separate `dbCall` method which throws this.

It's definitively possible (both cases) but both would require more code changes than the proposed change here.